src/adjustproxyconnection.cpp

```diff
@@ -0,0 +1,184 @@
+  MVPN_COUNT_DTOR(AdjustProxyConnection);
+  logger.debug() << "Connection released";
+}
+
+void AdjustProxyConnection::readData() {
+  Q_ASSERT(m_connection);
+  QByteArray input = m_connection->readAll();
+  m_buffer.append(input);
+
+  switch (m_state) {
+    case ProcessingState::NotStarted:
+      if (m_buffer.isEmpty()) {
+        return;
+      }
+      processFirstLine();
+    case ProcessingState::FirstLineDone:
+      if (m_buffer.isEmpty()) {
+        return;
+      }
+      processHeaders();
+    case ProcessingState::HeadersDone:
+      processParameters();
+    case ProcessingState::ParametersDone:
+      break;
+    default:
+      logger.debug() << "Unknown ProcessingState: " << m_state;
+  }
+
+  if (m_state == ProcessingState::ParametersDone) {
+    filterParameters();
+  }
+}
+
+void AdjustProxyConnection::processFirstLine() {
+  int pos = m_buffer.indexOf("\n");
+  if (pos == -1) {
+    return;
+  }
+
+  QByteArray line = m_buffer.left(pos);
+  m_buffer.remove(0, pos + 1);
+
+  QList<QByteArray> parts = line.split(' ');
+  if (parts.length() < 2) {
+    m_connection->close();
+    return;
+  }
+
+  m_method.append(parts[0]);
+  m_route = parts[1];
+
+  m_state = ProcessingState::FirstLineDone;
+}
+
+void AdjustProxyConnection::processHeaders() {
+  while (true) {
+    int pos = m_buffer.indexOf("\n");
+    if (pos == -1) {
+      break;
+    }
+
+    QByteArray line = m_buffer.left(pos);
+    m_buffer.remove(0, pos + 1);
+
+    QByteArray header = line.trimmed();
+    if (header == "") {
+      break;
+    }
+
+    pos = header.indexOf(":");
+    if (pos == -1) {
+      continue;
+    }
+
+    QByteArray headerName = header.left(pos);
+    header.remove(0, pos + 1);
+    QByteArray headerValue = header.trimmed();
+
+    QPair<QByteArray, QByteArray> headerPair;
+    headerPair.first = headerName;
+    headerPair.second = headerValue;
+
+    m_headers.append(headerPair);
+  }
+
+  m_state = ProcessingState::HeadersDone;
+}
+
+void AdjustProxyConnection::processParameters() {
+  if (m_method == "GET") {
+    m_paramters = QUrlQuery(m_route);
+  } else {
+    if (m_buffer.isEmpty()) {
+      return;
+    }
+    m_paramters = QUrlQuery(m_buffer.trimmed());
+  }
+
+  m_state = ProcessingState::ParametersDone;
+}
+
+void AdjustProxyConnection::filterParameters() {
+  QList<QPair<QString, QString>> newParameters;
+
+  for (QPair<QString, QString> parameter : m_paramters.queryItems()) {
+  }
+
+  forwardRequest();
+}
+
+void AdjustProxyConnection::forwardRequest() {
+  NetworkRequest* request;
+  if (m_method == "GET") {
+    m_route.setQuery(m_paramters);
+
+    request =
+        NetworkRequest::createForAdjustForwardGet(this, m_route, m_headers);
+  } else if (m_method == "POST") {
+    request = NetworkRequest::createForAdjustForwardPost(
+        this, m_route, m_headers, m_paramters.toString().toUtf8());
+  } else {
+    logger.warning() << "Method not supported!";
+    return;
+  }
+
+  connect(
```

Comment:
Maybe we should "proxy" the response entirely. Ignoring the request failure or success.

Reply:
For that, we would need to have access to the `m_reply` of the `NetworkRequest` which is private.

Reply:
we can expose it :)

Reply:
Yes, we can expose it. But then we would have to write the logic for all events for the reply because we could not use the `NetworkRequest`  implementation.
Since we don't proxy the request in guardian but create a new request with the same body and status I think this is not necessary.